### ISSUE TITLE
ENH: msg to ask who/what is running a plan

### ISSUE THREAD

**Issue by cjtitus** [bluesky/bluesky#1461]
## TLDR
We should provide a "which_engine" message that lets plans know who is running them. This would let adaptive plans define their own predictable execution during introspection and summarization. The return value of `Msg("which_engine")` needs discussion.

## Proposal and Justification

A core justification for the Bluesky msg/generator architecture is that it allows for plan introspection and simulation. Unfortunately, adaptive plans cannot be simulated, as they may not run correctly without being able to communicate with hardware. However, it would still be desirable to be able to introspect and simulate plans that contain adaptive sub-plans.

A new msg that would let a plan ask "who is running me" would provide a mechanism for adaptive plans to define their behavior during simulations, summaries, limit checks, etc. This could consist of skipping only the adaptive behavior, skipping the entire plan, or whatever else.

There are certain to be uses for a "who is running me" message beyond simulation of adaptive plans.

## Possible Implementations

The new command could be called, for example, "which_engine", which would have to return something to indicate what is running the plan.

`engine = Msg("which_engine")`

In discussion with @tacaswell the return value for the "which_engine" msg, was proposed to be either 

1. a descriptive string
2. the class of the engine
3. the actual instance of the engine 

`None` is always a possible return value. My gut instinct is that it would cause problems if there were a message that let a plan access the actual RunEngine instance. It's dangerous, unnecessary, and liable to be abused. 

Access to the engine class seems safer, and could allow for nice "automatic" implementations via type checking. However, there is currently only one RunEngine class, and none of the simulators are class-based, so this wouldn't be very useful yet.

string return values are easy to implement and interpret, and could be very easily put into the existing simulator code. However, it also seems messy to have every engine send its own strings, and now you have to keep track of all of the possible string values people have made up and implement checks for them. 

## Conclusion
I would like to have a discussion about whether we should return descriptive strings, or the engine class. If we return the class, should we also implement a `SimEngine` class to complement `RunEngine` with the idea that all the introspection tools in bluesky.simulators would subclass `SimEngine`? Then plans can easily check whether they are being run by the RunEngine, or some kind of simulator, and take appropriate action.

**Comment by untzag** [comment]
I'm surprised at the suggestion that simulations are done at the engine level. I've only thought about simulation implemented using fake devices and an untouched engine. Could device inspection be more appropriate here? One could imagine a simple `FakeDevice` mix-in class that does nothing but help plans do the right thing automatically.

Disclaimer: I have no deep knowledge of Bluesky :smile:

**Comment by cjtitus** [comment]
My suggestion wouldn't be to use the actual `RunEngine` for simulations, but to implement a separate `SimulationEngine` that just has a few convenience functions for handling messages. This a bit of a diversion from the main point however, as converting the current simulation _functions_ into simulation _classes_ is only relevant if we went with a class-based return value for "who is running me".

To be clear, when I say simulation, I mostly mean "introspection of a plan that you want to run with real hardware". So the `plot_raster_path` function is a "simulation", in that it is running through a plan, and inspecting the motor movements, but this is done by inspecting the messages, not by faking the devices. Technically, "summarize" is a simulation (at least it's in bluesky.simulators), but again, this isn't done via fake devices, it's just running through the generator messages and printing them rather than dispatching them to the real hardware.

I am leaning towards having the "who is running me" message return a string, which right now would either be "simulation" or "runengine", so that it is easy to keep simulators as simple functions, and we can leave the `SimulationEngine` concept for another day.

**Comment by untzag** [comment]
I see :smile:

**Comment by tacaswell** [comment]
xref that we already have a version of this :  https://github.com/bluesky/bluesky/pull/1299

**Comment by stan-dot** [comment]
@tacaswell  given that #1299 is merged can we close this?
